### PR TITLE
[Fix #4324] Add Lint/InterpolationCheck cop

### LINKED ISSUE #4324: Detect interpolation in ''-enclosed strings
It's easy to make a typing mistake like this:

```ruby
foo = 'something with #{interpolation} inside'
```

when you intended to write

```ruby
foo = "something with #{interpolation} inside"
```

This is almost always an error. It would be great if Rubocop could spot that, and help me find it faster (i.e. before I run my code and notice that it's b0rken. 😆)

## RuboCop version

```
$ rubocop -V
0.48.1 (using Parser 2.4.0.0, running on ruby 2.4.1 x86_64-darwin16)
```

### PR DESCRIPTION
**Pull request by GauthamGoli**
Adds the new cop as discussed in #4324 . I wrote the code, but there are many errors coming up which I've been trying to fix. Looking for help from maintainers/contributors so I can make further contributions without these hiccups. Please tell me what I'm doing wrong. Thanks!

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [ ] All tests are passing.
* [ ] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by pocke** [comment]
- When add a new cop, an entry in https://github.com/bbatsov/rubocop/blob/master/config/enabled.yml is required.
- Please try to run `bundle exec rake` in your computer. It executes `rspec` and `rubocop` for this project.

**Comment by pocke** [comment]
In spec of rubocop repository, the added cop has many false positives.
Because we use string interpolation in single quote for test. For example: https://github.com/bbatsov/rubocop/blob/00e1a2e5937dfa8f8146b2d75563edd70cdd1060/spec/rubocop/cop/security/eval_spec.rb#L55

So, maybe we should disable the cop in `spec/` directory by our project's `.rubocop.yml`. https://github.com/bbatsov/rubocop/blob/00e1a2e5937dfa8f8146b2d75563edd70cdd1060/.rubocop.yml


And  in `lib/` directory, the cop has several false positives. I guess the false positives should be disabled by inline comment. http://rubocop.readthedocs.io/en/latest/configuration/#disabling-cops-within-source-code

**Comment by pocke** [comment]
The cop raised some errors when I ran the cop in rubocop directory.

```bash
$ rubocop --only InterpolationCheck -d

(snip)

5 errors occurred:                                                    
An error occurred while Lint/InterpolationCheck cop was inspecting /home/pocke/ghq/github.com/bbatsov/rubocop/spec/rubocop/path_util_spec.rb:85:50.                                                                                                                                       
An error occurred while Lint/InterpolationCheck cop was inspecting /home/pocke/ghq/github.com/bbatsov/rubocop/spec/rubocop/result_cache_spec.rb:157:55.                                                                                                                                   
An error occurred while Lint/InterpolationCheck cop was inspecting /home/pocke/ghq/github.com/bbatsov/rubocop/spec/rubocop/cop/layout/end_of_line_spec.rb:16:33.                                                                                                                          
An error occurred while Lint/InterpolationCheck cop was inspecting /home/pocke/ghq/github.com/bbatsov/rubocop/spec/rubocop/cop/lint/syntax_spec.rb:50:0.                                                                                                                                  
An error occurred while Lint/InterpolationCheck cop was inspecting /home/pocke/ghq/github.com/bbatsov/rubocop/spec/rubocop/processed_source_spec.rb:80:8.  
```



It is a bug of the cop. Could you check the errors?
And if you execute your cop in large project(e.g. [gitlab](https://github.com/gitlabhq/gitlabhq), [redmine](https://github.com/redmine/redmine) and [ruby/spec](https://github.com/ruby/spec)), you may be able to find more errors if the errors exist.

File: spec/rubocop/cop/lint/interpolation_spec.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+describe RuboCop::Cop::Lint::InterpolationCheck do
+  subject(:cop) { described_class.new }
+
+  it 'registers an offense for interpolation in single quoted string' do
+    inspect_source(cop, <<-END.strip_indent)
```
Comment (Drenmi) [inline spec/rubocop/cop/lint/interpolation_spec.rb]: Please use the `expect_offense` matcher. 🙂

File: spec/rubocop/cop/lint/interpolation_spec.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+      'asdfasfasfds\#{testvar}alkdjfhaslkjhdfadslkjf'
+    END
+    expect(cop.offenses.size).to eq(1)
+  end
+
+  it 'does not register an offense for properly interpolation strings' do
+    expect_no_offenses(<<-END.strip_indent)
+      hello = "asdfasfasfds\#{testvar}alkdjfhaslkjhdfadslkjf"
```
Comment (Drenmi) [inline spec/rubocop/cop/lint/interpolation_spec.rb]: No need to make this almost impossible to read? 😅 

```ruby
hello = "foo #{bar}"
```

Comment (backus) [inline spec/rubocop/cop/lint/interpolation_spec.rb]: Also if you put single quotes around the heredoc then you don't need to escape the interpolation. Like `<<-'END'`

File: lib/rubocop/cop/lint/interpolated_string.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      #
+      #  foo = 'something with #{interpolation} inside'
+      #
+      # @example
+      #
+      #   # good
+      #
+      #  foo = "something with #{interpolation} inside"
+      class InterpolationCheck < Cop
+        MSG = 'Did you want to interpolate?'.freeze
```
Comment (Drenmi) [inline lib/rubocop/cop/lint/interpolated_string.rb]: This could be a bit less cryptic. 🙂

File: lib/rubocop/cop/lint/interpolated_string.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+        def on_str(node)
+          offense_source_range = regex_match_group(node, /(#\{.*\})/)
+          return unless offense_source_range
+          add_offense(node, offense_source_range, MSG)
```
Comment (Drenmi) [inline lib/rubocop/cop/lint/interpolated_string.rb]: `MSG` is redundant here. (RuboCop will tell you this if you're running latest `master` and run it.)

**Comment by Drenmi** [comment]
Looks like you're off to a good start! I left some small comments. Unfortunately, I don't have time to look at the highlighting code just now.

It's extremely common to get sucked down the rabbit hole, trying to add too many features at once, when first starting out, as evidenced by the many abandoned PRs. 🙂 Because of this I tend to recommend making a very simple version of the cop first (i.e. just mark the entire node as offense) and get that merged.

Then submit enhancements PRs for improvements (e.g. highlight only the interpolated part of the string.) 🙂

File: lib/rubocop/cop/lint/interpolated_string.rb (hunk 3 above)
Comment (GauthamGoli) [inline lib/rubocop/cop/lint/interpolated_string.rb]: Please suggest a better `MSG`

**Comment by GauthamGoli** [comment]
@pocke  Thanks! Make changes accordingly. 
@Drenmi Removed the code highlighting in this PR and will add it later!

File: config/enabled.yml [hunk 5]
```diff
@@ -1228,6 +1228,10 @@ Lint/InheritException:
   Description: 'Avoid inheriting from the `Exception` class.'
   Enabled: true
 
+Lint/InterplolationCheck:
```
Comment (pocke) [inline config/enabled.yml]: typo? `Interp**l**olationCheck -> InterpolationCheck`

File: lib/rubocop/cop/lint/interpolated_string.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+module RuboCop
+  module Cop
+    module Lint
+      # This cop checks for interpolation in a single quoted string.
+      #
+      # @example
+      #
+      #   # bad
+      #
+      #  foo = 'something with #{interpolation} inside'
```
Comment (Drenmi) [inline lib/rubocop/cop/lint/interpolated_string.rb]: Code examples need an additional space of indentation to be in line with `# bad` and `# good`. 😊

File: spec/rubocop/cop/lint/interpolation_spec.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  subject(:cop) { described_class.new }
+
+  it 'registers an offense for interpolation in single quoted string' do
+    expect_offense(<<-'RUBY'.strip_indent)
+      'foo #{bar}'
+      ^^^^^^^^^^^^ Did you want to interpolate?
+    RUBY
+  end
+
+  it 'does not register an offense for properly interpolation strings' do
```
Comment (Drenmi) [inline spec/rubocop/cop/lint/interpolation_spec.rb]: I would add a `no_offenses` test case for nested strings, i.e.:

```ruby
expect_no_offenses(<<-'RUBY'.strip_indent)
  foo = "bar '#{baz}' qux"
RUBY
```

🙂

File: lib/rubocop/cop/lint/interpolated_string.rb (hunk 3 above)
Comment (Drenmi) [inline lib/rubocop/cop/lint/interpolated_string.rb]: Hm. Maybe:

> Interpolation in single quoted string detected. Use double quoted strings if you need interpolation.

The first part could be removed when we add the improved highlighting. 🙂 WDYT?

File: config/enabled.yml (hunk 5 above)
Comment (Drenmi) [inline config/enabled.yml]: Once the typo is fixed, the CI should pass. 🙂

**Comment by Drenmi** [comment]
Just a few small fixes, then it's good! 🙂

**Comment by GauthamGoli** [comment]
@Drenmi @pocke Thanks again! Fixed them.

**Comment by Drenmi** [comment]
After thinking about it. This cop should probably not check regular heredocs. I think this is what's causing all the offenses on RuboCop itself. So instead of inline disabling, we should add a check for it.

```ruby
return if node.loc.is_a?(Parser::Source::Map::Heredoc)
```

**Comment by GauthamGoli** [comment]
@Drenmi I added the line, and removed inline disabling. But the offenses before inline disabling are coming up again.

**Comment by GauthamGoli** [comment]
@Drenmi Ping. I know this is late, but I haven't given up yet. Failed CI builds show some unrelated cops. Can you please take a look?

**Comment by schneems** [comment]
Is there a "blessed" way to indicate that no, i in-fact did not want to interpolate my string and it should actually have a `'#{pattern}`  in it?

**Pull request merged**